### PR TITLE
[new exo-run 4] boot up dependencies separately from services

### LINKED ISSUE #315: boot up dependencies separately from services
> When using dependencies that take a while to boot up
I want "exo run" to wait before it boots up services that depend on them
So that my services find fully running dependencies when they boot up and don't have to implement complicated wait or retry logic to connect to their dependencies.

- "exo run" first boots up all dependencies, and only when they all run then all the services
- the dependencies include application and service dependencies
- after the dependencies are up, it prints "dependencies are online"
- after the services are up, it prints "services are online"

### PR DESCRIPTION
**Pull request by cheahuychou**
resolves #315

### CODE REVIEW

File: exo-go/src/app_runner/index.go [hunk 1]
```diff
@@ -82,23 +89,49 @@ func (a *AppRunner) Shutdown(shutdownConfig types.ShutdownConfig) error {
...
+		return errors.Wrap(err, fmt.Sprintf("Failed to run services\nOutput: %s\nError: %s\n", process.Output, err))
+	}
+	for role, onlineText := range serviceOnlineTexts {
+		if err := a.waitForOnlineText(process, role, onlineText); err != nil {
+			return err
+		}
 	}
-	return err
+	a.write("all services online")
+	return nil
```
Comment (charlierudolph) [inline exo-go/src/app_runner/index.go]: `startDependencies` and `startServices`are very repetitive. I think the two could be dried up with something like:
```go
func startImages(imageNames []string, imageOnlineTexts map[string]string, identifier string) error {
  // identifier used when wrapping errors and for the final log message 
}
```

**Comment by cheahuychou** [comment]
@charlierudolph comment addressed

File: exo-go/src/docker_compose/index.go [hunk 2]
```diff
@@ -20,9 +20,9 @@ func PullAllImages(dockerComposeLocation string, log func(string)) error {
 	return processHelpers.RunAndLog(dockerComposeLocation, log, "docker-compose", "pull")
 }
 
-// RunAllImages runs all the docker images
-func RunAllImages(env []string, cwd string, log func(string)) (*processHelpers.Process, error) {
-	process := processHelpers.NewProcess("docker-compose", "up")
+// RunImages runs the given docker images
+func RunImages(images []string, env []string, cwd string, log func(string)) (*processHelpers.Process, error) {
+	process := processHelpers.NewProcess(append([]string{"docker-compose", "up"}, images...)...)
```
Comment (charlierudolph) [inline exo-go/src/docker_compose/index.go:25]: I think this line can be simplified to `processHelpers.NewProcess("docker-compose", "up", images...)`

**Review by charlierudolph — COMMENTED** [review]
Please merge with master as the waiting for online text changed with my process helpers PR. I want to take one more quick look after that is merged

File: exo-go/src/docker_compose/index.go (hunk 2 above)
Comment (cheahuychou) [reply exo-go/src/docker_compose/index.go:25]: I had already tried that. This was the error:
```
src/docker_compose/index.go:25: too many arguments in call to processHelpers.NewProcess
	have (string, string, []string...)
	want (...string)
```

Comment (charlierudolph) [reply exo-go/src/docker_compose/index.go:25]: Danget. Come on golang. Thats just weird

File: exo-go/src/app_runner/index.go [hunk 3]
```diff
@@ -63,6 +70,31 @@ func (a *AppRunner) getEnv() []string {
...
+		if err != nil {
+			return err
+		}
+		go func(role string, onlineTextRegex *regexp.Regexp) {
+			a.waitForOnlineText(process, role, onlineTextRegex)
+			wg.Done()
+		}(role, onlineTextRegex)
+	}
+	wg.Wait()
+	if err == nil {
```
Comment (charlierudolph) [inline exo-go/src/app_runner/index.go]: I know you were just copying code, but I believe this check can go away as err will never be non-nil at this point.

**Review by charlierudolph — APPROVED** [review]

**Comment by cheahuychou** [comment]
closed in: https://github.com/Originate/exosphere/commit/114650a41c58c210805ee4f2d9b3ae621ebff121

**Pull request closed without merging**